sys-utils/libcmt/src/io.c

```diff
@@ -0,0 +1,174 @@
+/* Copyright Cartesi and individual authors (see AUTHORS)
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "io.h"
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+#include <linux/cartesi/cmio.h>
+
+int cmt_io_init(cmt_io_driver_t *_me) {
+    int rc = 0;
+
+    if (!_me) {
+        return -EINVAL;
+    }
+    cmt_io_driver_ioctl_t *me = &_me->ioctl;
+    me->fd = open("/dev/cmio", O_RDWR);
+
+    if (me->fd < 0) {
+        rc = -errno;
+        return rc;
+    }
+
+    struct cmio_setup setup;
+    if (ioctl(me->fd, IOCTL_CMIO_SETUP, &setup)) {
+        rc = -errno;
+        goto do_close;
+    }
+
+    void *tx = mmap((void *) setup.tx.data, setup.tx.length, PROT_READ | PROT_WRITE, MAP_SHARED, me->fd, 0);
+    if (tx == MAP_FAILED) {
+        rc = -errno;
+        goto do_close;
+    }
+
+    void *rx = mmap((void *) setup.rx.data, setup.rx.length, PROT_READ, MAP_SHARED, me->fd, 0);
+    if (rx == MAP_FAILED) {
+        rc = -errno;
+        goto do_unmap;
+    }
+
+    me->rx_max_length = setup.rx.length;
+    me->rx_fromhost_length = 0;
+
+    cmt_buf_init(me->tx, setup.tx.length, tx);
+    cmt_buf_init(me->rx, setup.rx.length, rx);
+    return 0;
+
+do_unmap:
+    munmap(tx, setup.tx.length);
+do_close:
+    close(me->fd);
+    return rc;
+}
+
+void cmt_io_fini(cmt_io_driver_t *_me) {
+    if (!_me) {
+        return;
+    }
+    cmt_io_driver_ioctl_t *me = &_me->ioctl;
+
+    munmap(me->tx->begin, cmt_buf_length(me->tx));
+    munmap(me->rx->begin, cmt_buf_length(me->rx));
+    close(me->fd);
+
+    memset(me, 0, sizeof(*me));
+    me->fd = -1;
+}
+
+cmt_buf_t cmt_io_get_tx(cmt_io_driver_t *me) {
+    static const cmt_buf_t empty = {NULL, NULL};
+    if (!me) {
+        return empty;
+    }
+    return *me->ioctl.tx;
+}
+
+static uint32_t min(uint32_t a, uint32_t b) {
+    return a < b ? a : b;
+}
+
+cmt_buf_t cmt_io_get_rx(cmt_io_driver_t *me) {
+    static const cmt_buf_t empty = {NULL, NULL};
+    if (!me) {
+        return empty;
+    }
+    cmt_buf_t rx = *me->ioctl.rx;
+    rx.end = rx.begin + min(me->ioctl.rx_max_length, me->ioctl.rx_fromhost_length);
```

Comment:
`tx` and `rx` buffers are designed to mirror the kernel mmap buffers and thus be read only throughout the library use.
This makes it impossible for applications to know the available buffer size.

Reply:
Also, this file is designed to do the bare minimum abstraction of the kernel driver. This falls outside of that.